plugins/cli/__init__.py

```diff
@@ -22,234 +23,621 @@
 
-    def lo(self):
-        self.push("Logics:\n")
-        for logic in sorted(self.sh.return_logics()):
-            nt = self.sh.scheduler.return_next(logic)
-            if nt is not None:
-                self.push("{0} (scheduled for {1})\n".format(logic, nt.strftime('%Y-%m-%d %H:%M:%S%z')))
+    # noinspection PyUnusedLocal
+    def _cli_ld(self, handler, parameter, source):
+        if parameter is None or parameter == "":
+            log = self.sh.log
+        else:
+            logs = self.sh.return_logs()
+            if parameter not in logs:
+                handler.push("Log '{0}' does not exist\n".format(parameter))
+                log = None
             else:
-                self.push("{0}\n".format(logic))
+                log = logs[parameter]
 
-    def lt(self):
-        # list all threads with names
-        self.push("{0} Threads:\n".format(threading.activeCount()))
-        for t in threading.enumerate():
-            self.push("{0}\n".format(t.name))
-
-    def rt(self):
-        # return SH.py runtime
-        self.push("Runtime: {}\n".format(self.sh.runtime()))
-
-    def usage(self):
-        self.push('cl: clean (memory) log\n')
-        self.push('ls: list the first level items\n')
-        self.push('ls item: list item and every child item (with values)\n')
-        self.push('la: list all items (with values)\n')
-        self.push('lo: list all logics and next execution time\n')
-        self.push('lt: list current thread names\n')
-        self.push('update item = value: update the specified item with the specified value\n')
-        self.push('up: alias for update\n')
-        self.push('dump item: dump details about given item\n')
-        self.push('tr logic: trigger logic\n')
-        self.push('rl logic: reload logic\n')
-        self.push('rr logic: reload and run logic\n')
-        self.push('rt: return runtime\n')
-        self.push('quit: quit the session\n')
-        self.push('q: alias for quit\n')
-
-
-class CLI(lib.connection.Server):
-
-    def __init__(self, smarthome, update='False', ip='127.0.0.1', port=2323):
-        self.logger = logging.getLogger(__name__)
-        lib.connection.Server.__init__(self, ip, port)
-        self.sh = smarthome
-        self.updates_allowed = smarthome.string2bool(update)
-
-    def handle_connection(self):
-        sock, address = self.accept()
-        if sock is None:
-            return
-        self.logger.debug("{}: incoming connection from {} to {}".format(self._name, address, self.address))
-        CLIHandler(self.sh, sock, address, self.updates_allowed)
-
-    def run(self):
-        self.alive = True
-
-    def stop(self):
-        self.alive = False
-        self.close()
+        if log is not None:
+            handler.push("Log dump of '{0}':\n".format(log._name))
+            for entry in log.last(10):
+                values = [str(value) for value in entry]
+                handler.push(str(values))
+        handler.push("\n")
```

Comment:
The newline character should be on the end of each log line and not on the end of the whole output. So changing the indenting to 16 character would fix this output "error". :wink: